File: framework/core/compiler/main-compiler.js
import { compile as rustCompile } from './rust-compiler.js'
import { tokenize } from './lexer.js'
import { parse } from './parser.js'
import { transform } from './transformer.js'
import { generateSSR } from './ssr-generator.js'
import { CompileError } from './errors.js'

export function compile(source, options = {}) {
    const filename = options.filename ?? '<template>'
    const components = new Set(options.components ?? [])
    const target = options.target ?? 'client'
    let code = null
    let ssrCode = null
    let ast = null
    let tokens = null
    let warnings = []

    if (target === 'client' || target === 'both') {
        try {
            const result = rustCompile(source, {
                filename,
                runtimePath: options.runtimePath,
                components: [...components],
                scope: options.scope ? [...options.scope] : [],
                throwOnWarning: options.throwOnWarning ?? false,
            })
            code = result.code
            ast = result.ast
            tokens = result.tokens
            warnings = result.warnings ?? []
        } catch (err) {
            _rethrow(err, filename)
        }
    }

    if (target === 'ssr' || target === 'both') {
        try {
            let ssrTokens
            try {
                ssrTokens = tokenize(source)
            } catch (err) { _rethrow(err, filename) }

            let ssrAst
            try {
                ssrAst = parse(ssrTokens, source, components)
            } catch (err) { _rethrow(err, filename) }

            let ssrWarnings = []
            try {
                const result = transform(ssrAst, { components, source, scope: options.scope })
                ssrAst = result.ast
                ssrWarnings = result.warnings
            } catch (err) { _rethrow(err, filename) }

            if (options.throwOnWarning && ssrWarnings.length > 0) {
                const w = ssrWarnings[0]
                throw new CompileError(w.code, w.loc, source)
            }

            ssrCode = generateSSR(ssrAst, { source, runtimePath: options.ssrRuntimePath ?? options.runtimePath })

            if (target === 'ssr') {
                code = ssrCode
                tokens = ssrTokens
                ast = ssrAst
            }
            if (warnings.length === 0) {
                warnings = ssrWarnings
            }
        } catch (err) { _rethrow(err, filename) }
    }

    return {
        code,
        ssrCode,
        ast,
        tokens,
        warnings,
    }
}

export function parseOnly(source, options = {}) {
    const components = new Set(options.components ?? [])
    const tokens = tokenize(source)
    const ast = parse(tokens, source, components)
    return { ast, tokens }
}

export function transformOnly(ast, options = {}) {
    return transform(ast, {
        components: new Set(options.components ?? []),
        source: options.source ?? '',
        scope: options.scope,
    })
}

export function compileBatch(templates, sharedOptions = {}) {
    return templates.map(({ id, source }) => {
        try {
            const { code, warnings } = compile(source, { ...sharedOptions, filename: id })
            return { id, code, warnings, error: null }
        } catch (err) {
            return { id, code: null, warnings: [], error: err }
        }
    })
}

export function fuseePlugin(pluginOptions = {}) {
    return {
        name: 'vite-plugin-fusee',

        transform(src, id, ssrOptions) {
            if (!id.endsWith('.fusee') && !id.endsWith('.fhtml') && !id.endsWith('.template.html')) return null

            const isSSR = typeof ssrOptions === 'boolean' ? ssrOptions : Boolean(ssrOptions?.ssr)
            const { code, warnings } = compile(src, {
                filename: id,
                target: isSSR ? 'ssr' : 'client',
                runtimePath: pluginOptions.runtimePath,
                ssrRuntimePath: pluginOptions.ssrRuntimePath,
                components: pluginOptions.components,
                scope: pluginOptions.scope,
            })

            for (const w of warnings) {
                this.warn(typeof w.format === 'function' ? w.format(id) : String(w))
            }

            return { code, map: null }
        }
    }
}

export { fileRouterPlugin } from './plugins/file-router-plugin.js'
export { actionsPlugin } from './plugins/actions-plugin.js'
export { routerPlugin } from './plugins/router-plugin.js'
export { compileFileRoutes, generateRoutesModule, generateRouteTypes, validateRoutes as validateFileRoutes } from './plugins/file-router-plugin.js'
export { validateActions, generateClientStubs, generateServerRoutes, generateActionTypes } from './plugins/actions-plugin.js'
export { validateRoutes, generateRouterTypes } from './plugins/router-plugin.js'

function _rethrow(err, filename) {
    if (err instanceof CompileError) {
        err.message = err.format(filename)
        throw err
    }
    throw err
}

File: framework/core/rust-compiler/src/main_compiler.rs
use crate::ast::Node;
use crate::errors::{CompileError, CompileWarning};
use crate::generator::{generate, GenerateOptions};
use crate::lexer::{tokenize, Token};
use crate::parser::parse;
use crate::transformer::{transform, TransformOptions};
use std::collections::HashSet;

pub struct CompileOptions {
    pub filename: String,
    pub components: HashSet<String>,
    pub scope: HashSet<String>,
    pub runtime_path: Option<String>,
    pub throw_on_warning: bool,
}

impl Default for CompileOptions {
    fn default() -> Self {
        CompileOptions {
            filename: "<template>".to_string(),
            components: HashSet::new(),
            scope: HashSet::new(),
            runtime_path: None,
            throw_on_warning: false,
        }
    }
}

pub struct CompileResult {
    pub code: String,
    pub ast: Node,
    pub tokens: Vec<Token>,
    pub warnings: Vec<CompileWarning>,
}

pub fn compile(source: &str, options: CompileOptions) -> Result<CompileResult, CompileError> {
    let tokens = tokenize(source).map_err(|e| rethrow(e, &options.filename))?;

    let ast = parse(tokens.clone(), source, &options.components)
        .map_err(|e| rethrow(e, &options.filename))?;

    let transform_opts = TransformOptions {
        components: options.components.clone(),
        source: source.to_string(),
        scope: options.scope.clone(),
    };
    let (ast, warnings) = transform(ast, transform_opts);

    if options.throw_on_warning {
        if let Some(w) = warnings.first() {
            return Err(CompileError::new(w.code, w.loc.clone(), source, &[]));
        }
    }

    let root = match &ast {
        Node::Root(r) => r,
        _ => unreachable!("parse() always produces a Root node"),
    };

    let code = generate(
        root,
        GenerateOptions {
            source: source.to_string(),
            runtime_path: options.runtime_path.clone(),
        },
    )
        .map_err(|e| rethrow(e, &options.filename))?;

    Ok(CompileResult {
        code,
        ast,
        tokens,
        warnings,
    })
}

pub fn parse_only(source: &str, components: &HashSet<String>) -> Result<(Node, Vec<Token>), CompileError> {
    let tokens = tokenize(source)?;
    let ast = parse(tokens.clone(), source, components)?;
    Ok((ast, tokens))
}

pub fn transform_only(ast: Node, components: HashSet<String>, source: String, scope: HashSet<String>) -> (Node, Vec<CompileWarning>) {
    transform(
        ast,
        TransformOptions {
            components,
            source,
            scope,
        },
    )
}

fn rethrow(err: CompileError, filename: &str) -> CompileError {
    let formatted = err.format(filename);
    CompileError {
        message: formatted,
        ..err
    }
}